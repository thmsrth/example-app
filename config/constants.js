const APP_NAME = 'example';

const constants = {
  APP_NAME,
  MAIN_PATH: `/app/${APP_NAME}`,
};

module.exports = constants;
