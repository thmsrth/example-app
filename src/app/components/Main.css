.Main {
  position: relative;
}

.Main--error{
  box-shadow: var(--section-box-shadow);
  padding: var(--section-padding);
  margin: var(--section-margin);
  background-color: var(--section-background-color);
  border-top: var(--section-border-top);
  border-radius: var(--section-border-radius);
  text-align: center;
}

.Main__container--loading {
  opacity: 0.5;
}

.Main__actions {
  position: absolute;
  padding: var(--section-padding);
  text-align: center;
  width: 100%;
  margin: 3px 0;
  z-index: 99;
}

.Main__icon {
  margin-right: var(--space);
}

.Main__states {
  display: inline-flex;
  align-items: center;
}

.MainBack {
  padding: 0 var(--section-padding);
}

.MainBack__button__text,
.MainBack__button__icon {
  color: var(--color-black);
}

.Main__light {
  padding: calc(var(--space) / 10) calc(var(--space) / 4);
  border-radius: 10px;
  color: var(--color-white);
  font-size: 1.2rem;
  font-weight: 600;
}

.Main__light--green {
  background-color: var(--color-confirm);
}

.Main__light--red {
  background-color: var(--color-alert);
}

.Main__image--xs {
  height: auto;
  width: auto;
  max-height: 248px;
  max-width: 248px;
}
