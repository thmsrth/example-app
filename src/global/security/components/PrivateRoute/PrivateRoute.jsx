import React from 'react';
import PropTypes from 'prop-types';
import { Route, Redirect } from 'react-router-dom';

import { MAIN_PATH } from '../../../../../config/constants';

import DataStateNotifier from '../../../components/DataStateNotifier';

import SecurityUtil from '../../utils/SecurityUtil';

class PrivateRoute extends React.Component {
  componentDidMount() {
    const { checkAuthorization } = this.props;
    const authToken = SecurityUtil.getAuthToken();

    if (authToken) checkAuthorization(authToken);
  }

  render() {
    const {
      component: Component,
      permission,
      securityError,
      resetSecurityError,
      ...otherProps
    } = this.props;

    return (
      <DataStateNotifier
        dataError={securityError}
        resetDataError={resetSecurityError}
      >
        <Route
          {...otherProps}
          render={(props) => {
            if (SecurityUtil.isLoggedIn()) {
              if (SecurityUtil.hasAuthorized()) {
                return SecurityUtil.isGranted(permission)
                  ? <Component {...props} />
                  : <Redirect to={{ pathname: `${MAIN_PATH}/error403` }} />;
              }
              return <div />;
            }
            return <Redirect to={{ pathname: `${MAIN_PATH}/error401` }} />;
          }}
        />
      </DataStateNotifier>
    );
  }
}

export default PrivateRoute;

PrivateRoute.propTypes = {
  path: PropTypes.string,
  component: PropTypes.func,
  permission: PropTypes.string,
  securityError: PropTypes.string,
  resetSecurityError: PropTypes.func,
  checkAuthorization: PropTypes.func,
};

PrivateRoute.defaultProps = {
  path: '/',
  component: () => <div />,
  permission: '',
  securityError: '',
  resetSecurityError: () => {},
  checkAuthorization: () => {},
};
