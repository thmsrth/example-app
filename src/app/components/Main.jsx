import React from 'react';
import { Route, Switch } from 'react-router-dom';

import { MAIN_PATH } from '../../../config/constants';

import Register from '../../global/discovery/components/Register';
import Deregister from '../../global/discovery/components/Deregister';
import NotFound from '../../global/security/components/NotFound';
import NoAccess from '../../global/security/components/NoAccess';
import NotLoggedIn from '../../global/security/components/NotLoggedIn';
import PrivateRoute from '../../global/security/components/PrivateRoute';
import LoginPage from '../../global/security/components/LoginPage';

import MasterPage from './master/MasterPage';
import DetailPage from './detail/DetailPage';
import CreateExamplePage from './form/CreateExamplePage';
import EditExamplePage from './form/EditExamplePage';

import './Main.css';

const Main = props => (
  <div className="Main">
    <div className="Main__container">
      <Switch>
        <Route exact path={`${MAIN_PATH}/login`} component={LoginPage} />
        <Route exact path={`${MAIN_PATH}/error401`} component={NotLoggedIn} />
        <Route exact path={`${MAIN_PATH}/error403`} component={NoAccess} />

        <PrivateRoute
          {...props}
          exact
          path={`${MAIN_PATH}/create`}
          component={CreateExamplePage}
          permission="EXAMPLE_CREATE"
        />

        <PrivateRoute
          {...props}
          exact
          path={`${MAIN_PATH}/:id(\\d+)/edit`}
          component={EditExamplePage}
          permission="EXAMPLE_EDIT"
        />

        <PrivateRoute
          {...props}
          exact
          path={`${MAIN_PATH}/:id(\\d+)`}
          component={DetailPage}
          permission="EXAMPLE_DETAIL"
        />

        <PrivateRoute
          {...props}
          exact
          path={MAIN_PATH}
          component={MasterPage}
          permission="EXAMPLE_LIST"
        />

        <Route exact path="/register" component={Register} />
        <Route exact path="/deregister" component={Deregister} />

        <Route path="/" component={NotFound} />
      </Switch>
    </div>
  </div>
);

export default Main;
