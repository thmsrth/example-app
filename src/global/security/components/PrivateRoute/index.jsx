import React from 'react';
import { connect, Provider } from 'react-redux';

import securityStore from '../../store/security';

import {
  checkAuthorization,
  resetSecurityError,
} from '../../actions/security';

import {
  getPermissions,
  getSecurityError,
  getAuthorizing,
} from '../../reducers/security';

import PrivateRoute from './PrivateRoute';

const mapStateToProps = state => ({
  permissions: getPermissions(state),
  securityError: getSecurityError(state),
  authorizing: getAuthorizing(state),
});

const mapDispatchToProps = dispatch => ({
  checkAuthorization: params => dispatch(checkAuthorization(params)),
  resetSecurityError: () => dispatch(resetSecurityError()),
});

const ConnectedContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(PrivateRoute);

const PrivateRouteContainer = props => (
  <Provider store={securityStore}>
    <ConnectedContainer {...props} />
  </Provider>
);

export default PrivateRouteContainer;
