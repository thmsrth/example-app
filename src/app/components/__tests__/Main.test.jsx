import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { MAIN_PATH } from '../../../../config/constants';

import PrivateRoute from '../../../global/security/components/PrivateRoute';

import Main from '../Main';

global.authToken = 'jhHGFjHtfjhgfK';
global.permissions = {
  EXAMPLE_CREATE: true,
  EXAMPLE_EDIT: true,
  EXAMPLE_DETAIL: true,
  EXAMPLE_LIST: true,
};

const props = {
  securityError: '',
};

describe('Main', () => {
  it('should render the routes correctly', () => {
    const tree = shallow(<Main {...props} />);
    const paths = tree.find(PrivateRoute).reduce((map, route) => {
      const pathMap = { ...map };
      const componentProp = route.props().component;
      if (componentProp) {
        // The component is passed to Route as a component prop
        pathMap[route.prop('path')] = componentProp.name;
      } else {
        // The component is passed to Route as a render prop with routerProps as first argument
        const routerProps = { match: { params: { id: 1 } } };
        const renderProp = route.props().render(routerProps);
        pathMap[route.prop('path')] = renderProp.type.name;
      }

      return pathMap;
    }, {});
    expect(paths[MAIN_PATH]).toBe('MasterPage');
    expect(paths[`${MAIN_PATH}/:id(\\d+)`]).toBe('DetailPage');
    expect(paths[`${MAIN_PATH}/:id(\\d+)/edit`]).toBe('EditExamplePage');
    expect(paths[`${MAIN_PATH}/create`]).toBe('CreateExamplePage');
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render a security error message', () => {
    const tree = shallow(<Main
      {...props}
      securityError="My Security Error Message"
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
