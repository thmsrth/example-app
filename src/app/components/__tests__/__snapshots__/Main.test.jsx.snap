// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render a security error message 1`] = `
<div
  className="Main"
>
  <div
    className="Main__container"
  >
    <Switch>
      <Route
        component={[Function]}
        exact={true}
        path="/app/example/login"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/app/example/error401"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/app/example/error403"
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example/create"
        permission="EXAMPLE_CREATE"
        securityError="My Security Error Message"
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example/:id(\\\\d+)/edit"
        permission="EXAMPLE_EDIT"
        securityError="My Security Error Message"
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example/:id(\\\\d+)"
        permission="EXAMPLE_DETAIL"
        securityError="My Security Error Message"
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example"
        permission="EXAMPLE_LIST"
        securityError="My Security Error Message"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/register"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/deregister"
      />
      <Route
        component={[Function]}
        path="/"
      />
    </Switch>
  </div>
</div>
`;

exports[`Main should render the routes correctly 1`] = `
<div
  className="Main"
>
  <div
    className="Main__container"
  >
    <Switch>
      <Route
        component={[Function]}
        exact={true}
        path="/app/example/login"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/app/example/error401"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/app/example/error403"
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example/create"
        permission="EXAMPLE_CREATE"
        securityError=""
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example/:id(\\\\d+)/edit"
        permission="EXAMPLE_EDIT"
        securityError=""
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example/:id(\\\\d+)"
        permission="EXAMPLE_DETAIL"
        securityError=""
      />
      <PrivateRouteContainer
        component={[Function]}
        exact={true}
        path="/app/example"
        permission="EXAMPLE_LIST"
        securityError=""
      />
      <Route
        component={[Function]}
        exact={true}
        path="/register"
      />
      <Route
        component={[Function]}
        exact={true}
        path="/deregister"
      />
      <Route
        component={[Function]}
        path="/"
      />
    </Switch>
  </div>
</div>
`;
